src/Plugin.hs

```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE GADTs #-}
+
+module Plugin (plugin) where
+
+import Control.Concurrent.STM
+import Control.Exception
+import Control.Monad
+import Data.IORef
+import GHC
+import GHC.Driver.Hooks
+import GHC.Driver.Pipeline
+import GHC.Driver.Pipeline.Phases
+import GHC.Plugins as Plugins
+import GHC.Types.Name.Cache
+import HieDb.Create
+import HieDb.Types
+import System.Directory (doesPathExist)
+import System.FilePath
+import qualified System.IO.Unsafe as Unsafe
+
+plugin :: Plugin
+plugin =
+  defaultPlugin
+    { pluginRecompile = Plugins.purePlugin
+    , driverPlugin = driver
+    }
+
+driver :: [CommandLineOption] -> HscEnv -> IO HscEnv
+driver _ hscEnv = do
+  initializeHiedb
+  pure
+    hscEnv
+      { hsc_hooks =
+          (hsc_hooks hscEnv)
+            { runPhaseHook = Just phaseHook
+            }
+      }
+ where
+  initializeHiedb = liftIO $ withHieDb defaultHiedbFile initConn
+
+  -- We index using a phase hook instead of typeCheckResultAction since
+  -- the hie file can be written after that plugin phase
+  phaseHook =
+    PhaseHook $ \phase -> do
+      case phase of
+        T_HscPostTc _ modSummary _ _ _ -> do
+          let dynFlags = hsc_dflags hscEnv
+              hieDirectory = hieDir dynFlags
+          _ <- liftIO $ addModuleToDb defaultHiedbFile (ms_mod modSummary) hieDirectory
+          runPhase phase
+        _ -> runPhase phase
+
+addModuleToDb :: FilePath -> Module -> Maybe FilePath -> IO ()
+addModuleToDb hiedbFile mod' mHieBaseDir = do
+  let
+    -- Note: For performance reasons we intentionally skip the type
+    -- indexing phase
+    -- TODO: pass this in as a user defined option
+    skipOptions = defaultSkipOptions{skipTypes = True}
+    modToPath = moduleNameSlashes . moduleName
+
+  let mHieFile = ((</>) <$> mHieBaseDir) <*> pure (modToPath mod' -<.> ".hie")
+
+  case mHieFile of
+    Nothing -> pure ()
+    Just hieFile -> do
+      hieExists <- doesPathExist hieFile
+      when hieExists $ do
+        _ <- acquireDbLock
+        nc <- newIORef =<< initNameCache 'a' []
+        _ <-
+          withHieDb
+            hiedbFile
+            (\conn -> runDbM nc $ addRefsFrom conn (Just ".") skipOptions hieFile)
+            -- If indexing fails we dont want to
+            -- TODO: report this and maybe make configurable in future versions
+            `catch` (\(_ :: SomeException) -> pure False)
+        _ <- releaseDbLock
+        pure ()
+ where
+  acquireDbLock =
+    liftIO $ atomically $ takeTMVar dbLock
+  releaseDbLock =
+    liftIO $ atomically $ putTMVar dbLock ()
```

Comment:
maybe combine these into a `withDbLock :: IO a -> IO ` so that exceptions are handled correctly

Reply:
Added!